### PR TITLE
Improve handling of closed and re-opened JAR classpath files

### PR DESCRIPTION
**Pull request by kriegaex**
`ClasspathJar.openArchives` needs more safeguards when accessing possibly closed zip files (actually classpath JARs), making sure the get re-opened by calling `ClasspathJar.ensureOpen()`.

Relates to https://github.com/eclipse/org.aspectj/issues/125.

### CODE REVIEW

File: org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java [hunk 1]
```diff
@@ -202,18 +233,14 @@ public boolean hasAnnotationFileFor(String qualifiedTypeName) {
...
-        ensureOpen();
-	} catch (IOException ioe) {
-		// Doesn't normally occur - probably means since starting the compile
-		// you have removed one of the jars.
-		ioe.printStackTrace();
-		return null;
+		ensureOpen();
+	}
+	catch (IOException e) {
+		throw new RuntimeException(e);
```
Comment (aclement) [inline org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: This is a change in behaviour. I'm not sure we have the tests to validate making this change, do we?

File: org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java [hunk 2]
```diff
@@ -285,15 +312,12 @@ protected void addToPackageCache(String fileName, boolean endsWithSep) {
...
-        ensureOpen();
-	} catch (IOException ioe) {
-		// Doesn't normally occur - probably means since starting the compile
-		// you have removed one of the jars.
-		ioe.printStackTrace();
-		return singletonModuleNameIf(false);
+		ensureOpen();
+	}
+	catch (IOException e) {
+		throw new RuntimeException(e);
```
Comment (aclement) [inline org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:320]: this one too is a change in behaviour isn't it?

File: org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java (hunk 1 above)
Comment (kriegaex) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: Yes, it is a change in behaviour. In this situation, IMO not just nothing should happen but an error should be thrown. Why would we just return `false` where it is clear that ECJ expects an open zip for scan for entries instead? We cannot just overrule that because we messed up our own open classpath JAR cache. The error, if it ever happens should escalate up, so users can report it and we can fix it. That is a good thing, just like the error we are fixing in this PR.

File: org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java (hunk 2 above)
Comment (kriegaex) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:320]: Yes. Same rationale as above.

File: org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java (hunk 1 above)
Comment (kriegaex) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: Please also note the related new test in AspectJ. I have no idea if it covers all execution paths, but at least it checks that the problem addressed in this PR is fixed and will detect regressions in the future. That is one more test than we had before, and FWIW the other existing ones are still passing.

Comment (aclement) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: personally, due to bandwidth concerns, I would only fix what is broken according to the issue. Because I may not have time to fix whatever situation could occur due to the change in behaviour. If you feel you will have the time, maybe we can risk it. Sometimes the work necessary to be done to bend JDT behaviour to the needs of AspectJ is so massive that it isn't worth the time investment when the alternative is so simple (see many cases of how ITDs are handled).

Comment (kriegaex) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: FYI, I am currently running a CI build globally setting `-Dorg.aspectj.weaver.openarchives=20` for **all tests**. Feel free to watch [this build](https://github.com/kriegaex/org.aspectj/actions/runs/1890493915). The corresponding change in a private branch is https://github.com/kriegaex/org.aspectj/commit/eab77587e6916bd994f1b56beb2736583aefa838. I.e., we have hundreds of compilations, all running into JAR cache exhaustion due to the low limit. This should be a good indicator of what I potentially broke or did not fix yet.

Comment (kriegaex) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: > due to bandwidth concerns, I would only fix what is broken according to the issue.

I understand your concern. I think that I am able to revert those changes, if necessary in the future. I simply applied the boyscout rule: Leave the camp ground behind a little bit cleaner than you met it.

Comment (kriegaex) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: Good news: The stress-testing build for the JAR cache passed except [one flaky performance test on JDK 14](https://github.com/kriegaex/org.aspectj/runs/5312395732?check_suite_focus=true#step:6:17025). I also scanned all three build logs for `ensureOpen` and did not get any hits. so it seems that we do not catch and log corresponding exceptions anywhere. I am quite confident now that the changes are OK.

Comment (aclement) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: The comment in the code suggests maybe to me this is an AJDT occurrence (class path changing after build running - is it related to an incremental build after a full build but the class path is shorter?). It would be interesting to check the change history on that comment - did it used to throw the exception like elsewhere but then get changed, did somebody leave a sensible comment on the change? (my hopes are not too high though!). 

I do admire that you want to make the code better, I'm just a little nervous.  Can the null path you are removing lead to zip file being null (and so in fact be the problem this person is seeing? I didn't review that closely). I would have perhaps hoped he'd have reported the stack trace that was being dumped out in that case but maybe it was swallowed and he didn't see it.

Comment (kriegaex) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: You added that two-line comment in March 2014 in [this commit](https://git.eclipse.org/c/aspectj/org.aspectj.shadows.git/commit/org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java?h=aj_19&id=2c4c1f50c90225ec2b302a80d1c68aeca1ec2c04). Unfortunately, I had to go and find it in the old "shadows" repository. Git Blame display is broken on that GitLab instance, the commit ID column does not match the file on the right. But I have a local copy and found the change on the CLI. Maybe I just push another copy to my GitHub account, because the Eclipse admins removed all mirror repositories a while ago, remember?

Comment (kriegaex) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: BTW, the [workflow re-run](https://github.com/kriegaex/org.aspectj/actions/runs/1890493915) with cache size 20 passed.

Comment (kriegaex) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: And here is the [Git Blame of this class on GitHub](https://github.com/kriegaex/org.aspectj.shadows/blame/9dc0d904ee04d79dc40d2fd7b758c7b9e8a4b822/org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java#L204) - so much nicer to navigate than Eclipse GitLab.

Comment (aclement) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: I was thinking someone *might* be able to recreate those paths that might lead to null with a MultiProjectIncremental test (using that AspectJ infra) that did a full build with a crafted class path containing an extra entry and a low openArchive count, then deleted that special entry from the filesystem then do an incremental build that attempted to find something in it. Not trivial though.

Comment (kriegaex) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: Before I consider writing such a test, would you mind explaining which real-world usage scenario this ought to reflect? Can it be achieved in any correct interaction in Eclipse IDE? If so, please describe it. Otherwise, why would incremental compilation need to succeed and silently accept that situation instead of failing and forcing the user to do a full recompile, which IMO would be the right thing to do if a JAR containing necessary classes is no longer present on the file system. I think that JDT Core is not designed to consider such a scenario. Probably I am misunderstanding your scenario description somehow, but the way I (mis?)understand it, it simply makes no sense to me.

I think the old comment is not about deleting something from the file system but deleting something from the classpath in an inconsistent way (i.e. a possible bug), forgetting to also delete the file in the open JAR cache. That case should fail during compilation, IMO.

Comment (aclement) [reply org.eclipse.jdt.core/batch/org/eclipse/jdt/internal/compiler/batch/ClasspathJar.java:241]: you might be right, I haven't had a lot of time to dig into the code but it does occur to me if a class path entry was backed by a zip file that got closed, which would then allow it to be deleted, then it attempted to be used again we would probably fail. Doing something different in that try catch block would be, I'm 99.9% sure, because the situation happened in development/testing.

If I were in AJDT, a full build happened, I removed a bunch of code using a particular class, I then deleted a file off the disk (Eclipse won't notice until you refresh). I then added a new class and as part of searching from references from that file whilst during an incremental compilation it still looked in the class path entry that no longer was able to open the associated zip file, things would fail. (How you can treat these files - can you delete them, rename them - with the code opening and closing them can vary by Operating System).

> I think the old comment is not about deleting something from the file system but deleting something from the classpath in an inconsistent way 
if it were deleted from the class path, I don't know why we would be back in the object representing the class path entry trying to open the zip?

Is it worth the effort of writing a test? well I wouldn't do it as it is a corner case and if I had time to do this, it's better spent elsewhere. I think I'm just going to merge it and move on.
